[mlir] Rewrites for I2 to I8 signed and unsigned extension

ziereis opened the pull request:
Adds rewrites for i2 to i8 signed and unsigned extension, similar to the ones that already exist for i4 to i8 conversion.

I use this for i6 quantized models, and this gives me roughly a 2x speedup for an i6 4096x4096 dequantization-matmul on an AMD 5950x.

I didn't add the rewrite for i8 to i2 truncation because I currently don't use it, but if this is needed, I can add it as well.

ziereis committed:
init

ziereis committed:
fix typos in comments

ziereis committed:
fix typos in comments

banach-space commented on mlir/test/Dialect/Vector/vector-rewrite-narrow-types.mlir:
```diff
@@ -234,6 +234,72 @@ func.func @aligned_extsi_2d(%a: vector<8x32xi4>) -> vector<8x32xi32> {
 }
 
+// CHECK-LABEL: func.func @aligned_extsi_i2_to_i8(
+func.func @aligned_extsi_i2_to_i8(%a: vector<8xi2>) -> vector<8xi8> {
```
[nit] I'd be tempted to move `@aligned_extsi_i2_to_i8` next to `@aligned_extsi_i4_to_i8`. Same for other tests.

banach-space commented on mlir/lib/Dialect/Vector/Transforms/VectorEmulateNarrowType.cpp:
```diff
@@ -1233,6 +1233,117 @@ static Value rewriteI4ToI8UnsignedExt(PatternRewriter &rewriter, Location loc,
+  Value shifted3 = rewriter.create<arith::ShRUIOp>(loc, i8Vector, shiftValues3);
+  Value elem3 = rewriter.create<arith::AndIOp>(loc, shifted3, maskValues);
+
+  // interleave all 4 elements by first interleaving even elements and then odd
```
```suggestion
  // Interleave all 4 elements by first interleaving even elements and then odd.
```

banach-space commented on mlir/lib/Dialect/Vector/Transforms/VectorEmulateNarrowType.cpp:
```diff
@@ -1233,6 +1233,117 @@ static Value rewriteI4ToI8UnsignedExt(PatternRewriter &rewriter, Location loc,
+  Value elem0 = rewriter.create<arith::AndIOp>(loc, i8Vector, maskValues);
+
+  // Element 1 (bits 2-3)
+  constexpr int8_t shift1 = 2;
+  auto shiftAttr1 = DenseElementsAttr::get(i8VecType, shift1);
+  auto shiftValues1 = rewriter.create<arith::ConstantOp>(loc, shiftAttr1);
+  Value shifted1 = rewriter.create<arith::ShRUIOp>(loc, i8Vector, shiftValues1);
+  Value elem1 = rewriter.create<arith::AndIOp>(loc, shifted1, maskValues);
```
This block is repeated multiple times. Also, similar blocks are present in other functions. 

Could you try introducing an utility function instead? For example, `Value extractNBitsFromVal(int bitIdxStart, int bitIdxEnd, Value src)`?

ziereis replied:
Very good idea, thank you. I created the function with ```int bitIdx, int numBits``` instead of a start and end index because this was easier for me to reason about.

banach-space commented on mlir/lib/Dialect/Vector/Transforms/VectorEmulateNarrowType.cpp:
```diff
@@ -1233,6 +1233,117 @@ static Value rewriteI4ToI8UnsignedExt(PatternRewriter &rewriter, Location loc,
   return rewriter.create<vector::InterleaveOp>(loc, low, high);
 }
 
+/// Rewrite the i2 -> i8 signed extension into a sequence of shuffles and
+/// bitwise ops that take advantage of high-level information to avoid leaving
+/// LLVM to scramble with peephole optimizations.
```
> take advantage of high-level information

There's not much "high-level" info in arith.extsi ;-)

>  avoid leaving  LLVM to scramble with peephole optimizations

IMO, this is a perfect opportunity for the LLVM backend to shine. Once LLVM's optimizer can handle this, other frontends (i.e., not just MLIR + Vector) will benefit as well.

Just to clarify, I'm not against having these patterns here - quite the opposite! I'm very thankful that you're adding them. To me, this is more about working around LLVM's current limitations and filling the gap while LLVM catches up.

ziereis replied:
Makes sense. Honestly, I just copied these comments from the other functions :D. I removed the part about the "high-level" information in the comments.

banach-space reviewed (commented):
Thanks for working on this!

> I didn't add the rewrite for i8 to i2 truncation because I currently don't use it, but if this is needed, I can add it as well.

A TODO + negative test would be sufficient for me.

ziereis committed:
refactoring:
- extracts repeated code into functions
- reorder tests
- improve naming

ziereis commented on mlir/lib/Dialect/Vector/Transforms/VectorEmulateNarrowType.cpp:
```diff
@@ -1489,6 +1610,10 @@ struct RewriteAlignedSubByteIntTrunc : OpRewritePattern<arith::TruncIOp> {
                                              truncOp)))
       return failure();
 
+    // not supported currently.
```
Done. Also, a negative test should exist. I just check that I don't match on an i2 to i8 truncation, is that not sufficient?

ziereis commented:
Thanks a lot for the review. I believe I have addressed all the comments. I also noticed a superfluous ```AndIOp`` for the unsigned case, so I removed it and adjusted the tests accordingly.

hanhanW commented:
cc @lialan who recently works on this area.

hanhanW commented on mlir/lib/Dialect/Vector/Transforms/VectorEmulateNarrowType.cpp:
```diff
@@ -1084,8 +1084,8 @@ static LogicalResult alignedConversionPrecondition(PatternRewriter &rewriter,
 
-  // Only {s}i4 -> (size_of({{s}i/f}) >= 8) are supported for now.
-  if (srcElemBitwidth != 4 || dstElemBitwidth < 8 ||
+  // Only {s}i4/i2 -> (size_of({{s}i/f}) >= 8) are supported for now.
+  if ((srcElemBitwidth != 4 && srcElemBitwidth != 2) || dstElemBitwidth < 8 ||
       (dstElemBitwidth % srcElemBitwidth) != 0)
     return rewriter.notifyMatchFailure(op, "Not a supported aligned case");
```
This statement is becoming more complicated, and it is not easy to parse if the comment match the check or not. How about we break it into three if-statement and collapse the comment into the error message? E.g.,

```cpp
// I feel that it is a requirement for emulation, so we can just say that it must be greater than or equal to 8
if (dstElemBitwidth < 8)
  return rewriter.notifyMatchFailure(op, "the bitwidth of dstType must be greater than or equal to 8");
if (dstElemBitwidth % srcElemBitwidth != 0)
  return ... "unaligned cases are not supported"
if (srcElemBitwidth != 2 && srcElemBitwidth != 4)
  return ... "only support src bitwidth to be 2 or 4 at this moment"
}
```

hanhanW commented on mlir/lib/Dialect/Vector/Transforms/VectorEmulateNarrowType.cpp:
```diff
@@ -1172,70 +1172,167 @@ Value BitCastRewriter::genericRewriteStep(
   return runningResult;
 }
 
+Value bitcastSubByteVectorToI8(PatternRewriter &rewriter, Location loc,
+                               Value srcValue) {
+  VectorType srcVecType = cast<VectorType>(srcValue.getType());
+  int64_t srcBitwidth = srcVecType.getElementType().getIntOrFloatBitWidth();
+  assert(srcBitwidth % 8 != 0 && "Invalid source bitwidth");
+  int64_t bitwidthFactor = 8 / srcBitwidth;
+  SmallVector<int64_t> i8VecShape = llvm::to_vector(srcVecType.getShape());
+  i8VecShape.back() = i8VecShape.back() / bitwidthFactor;
+  auto i8VecType = VectorType::get(i8VecShape, rewriter.getI8Type());
+  return rewriter.create<vector::BitCastOp>(loc, i8VecType, srcValue);
+}
```
Few suggestions:

- Add a documentation to the function.
- Add `static` keyword to the local function. Same for the below function. https://llvm.org/docs/CodingStandards.html#anonymous-namespaces
- Use auto for `srcVecType` because `cast` already spells the type. https://llvm.org/docs/CodingStandards.html#use-auto-type-deduction-to-make-code-more-readable
- [optional] I'd rename `i8VecShape` to `vecShape`.
- You do not need the llvm::to_vector. One of the SmallVector constructors takes ArrayRef type directly. So it can be `SmallVector<int64_t> vecShape(srcVecType.getShape())`.

hanhanW commented on mlir/lib/Dialect/Vector/Transforms/VectorEmulateNarrowType.cpp:
```diff
@@ -1172,70 +1172,167 @@ Value BitCastRewriter::genericRewriteStep(
+
+/// Extracts a signed N-bit sequence from each element of an 8-bit vector,
+/// starting at the specified bit index.
+Value extractNBitsFromVectorSigned(PatternRewriter &rewriter, Location loc,
```
nit: add `static` keyword to the local function.
[optional]: it looks better if you can have an example in the doc.

hanhanW commented on mlir/lib/Dialect/Vector/Transforms/VectorEmulateNarrowType.cpp:
```diff
@@ -1172,70 +1172,167 @@ Value BitCastRewriter::genericRewriteStep(
+
+/// Extracts an unsigned N-bit sequence from each element of an 8-bit vector,
+/// starting at the specified bit index.
+Value extractNBitsFromVectorUnsinged(PatternRewriter &rewriter, Location loc,
```
nit: add `static` keyword

hanhanW commented on mlir/lib/Dialect/Vector/Transforms/VectorEmulateNarrowType.cpp:
```diff
@@ -1438,11 +1535,31 @@ struct RewriteAlignedSubByteIntExt : OpRewritePattern<ConversionOpType> {
     // Perform the rewrite.
     Value subByteExt;
     if (isSigned) {
```
nit: `conversionOp.getLoc()` is used several times. I think we can declare a variable, which makes code simpler. E.g., `Location loc = conversionOp.getLoc();`

hanhanW reviewed (changes requested):
The logic of signed and unsigned conversion is very similar. The only difference is that they use different `extractNBits*` functions. I think we can pass the function to the method, so we no longer need to duplicate the logic. E.g.,

```
using ExtractNBitsFn = std::function<Value(PatternRewriter&, Location, Value, int, int)>;

static Value rewriteI4ToI8(PatternRewriter &rewriter, Location loc,
                                    Value srcValue, ExtractNBitsFn extFn) {
// ...
  Value low = extFn(rewriter, loc, i8Vector, 0, 4);
  Value high = extFn(rewriter, loc, i8Vector, 4, 4);
// ...
}
```

ziereis committed:
refactoring

ziereis commented:
Thanks again for the review. All comments should be addressed.

banach-space commented on mlir/lib/Dialect/Vector/Transforms/VectorEmulateNarrowType.cpp:
```diff
@@ -1172,70 +1176,144 @@ Value BitCastRewriter::genericRewriteStep(
+  Value i8Vector = bitcastSubByteVectorToI8(rewriter, loc, srcValue);
+
+  // 2. Extract each i2 element
+  // Element 0 (bits 0-1)
+  Value elem0 = extFn(rewriter, loc, i8Vector, 0, 2);
+  // Element 1 (bits 2-3)
+  Value elem1 = extFn(rewriter, loc, i8Vector, 2, 2);
+  // Element 2 (bits 4-5)
+  Value elem2 = extFn(rewriter, loc, i8Vector, 4, 2);
+  // Element 3 (bits 6-7)
+  Value elem3 = extFn(rewriter, loc, i8Vector, 6, 2);
```
Are these vectors or scalars?

ziereis replied:
vectors, i fixed the naming

banach-space commented on mlir/lib/Dialect/Vector/Transforms/VectorEmulateNarrowType.cpp:
```diff
@@ -1172,70 +1176,144 @@ Value BitCastRewriter::genericRewriteStep(
+
+  // 3. Interleave all 4 elements by first interleaving
+  // even elements and then odd
+  // elem0 = [0,0,0,0]
+  // elem1 = [1,1,1,1]
+  // elem2 = [2,2,2,2]
+  // elem3 = [3,3,3,3]
```
If `elem0` is a scalar byte, then I read that as: 4 x i2, where every value is `0` (or `1` etc). But are these values really replicated like this?

ziereis replied:
elem0 is a vector, this is bad naming sorry. The values are just  "some values from vector i"

banach-space reviewed (commented):
Thanks for all the updates so far!

ziereis commented on mlir/lib/Dialect/Vector/Transforms/VectorEmulateNarrowType.cpp:
```diff
@@ -1172,70 +1176,144 @@ Value BitCastRewriter::genericRewriteStep(
+/// extract numBits=2 starting at bitIdx=2
+/// src                 = [0101|10|10]
+/// mask                = [00000011]
+/// shr    = src >> 6   = [00010110]
+/// result = shr & mask = [00000010]
```
it would, but you would end up with more instructions.

Example for 4 bit:

Shifts:
src : i8
shl = arith::shl(src)
low = artih::shrui(shl)
high = arith::shrui(src)

Mask:
src: i8
low = mask(src, 15)
high = artih::shrui(src)

banach-space replied:
In ideal world, we'd keep `extractNBitsFromVectorSigned` and `extractNBitsFromVectorUnsinged` very consistent, so I'm trying to understand the reason for divergence.

> Example for 4 bit:
> 
> Shifts: src : i8 
> shl = arith::shl(src) 
> low = artih::shrui(shl) 
> high = arith::shrui(src)

Let me make sure I got this right. I assume that we are extracting 2 bits at index 0 (indexing from MSB, i.e. left-most bit):
```bash
src: [10 10 11 00]
shl: arith.shl(src) --> [10 10 11 00]
shr: arith.shrui(shl) --> [00 00 00 10]
```
Why do we need another `shr` in your example? (i.e., what's `low` and `high`?)

ziereis replied:
okay, so my understanding is that when talking about bits the convention is to "read it right to left", so the LSB is at index 0. So the "low bits" are the right most bits and the "high bits" are the left most. 

src:         [ 1 | 0 | 1 | 0 | 1 | 1 | 0 | 0 ]
indices:  [ 7 | 6 | 5 | 4 | 3 | 2 | 1 | 0 ]

So i could always extract N bits starting from 0 with a single arith::AndIOp + mask. 
If i would use the shl + shrui pattern i would need to use 2 instructions. 

So this is basically a small shortcut i can take with unsigned numbers, but for the signed case i have to always do it with the shrsi for the sign.

banach-space replied:
> okay, so my understanding is that when talking about bits the convention is to "read it right to left", so the LSB is at index 0. So the "low bits" are the right most bits and the "high bits" are the left most. 

Yes, that makes more sense 😅  But then, when storing a _vector_ of 8 elements in memory, it's usually like: `v[0], v[1], v[2], ...`. Not to mention the split into little-endian/big-endian (but we are not storing anything in memory here, so not relevant). My point being - this is worth documenting.

> I will add this to the documentation

That was going to be my next request 😅 

> So this is basically a small shortcut i can take with unsigned numbers, but for the signed case i have to always do it with the shrsi for the sign.

TBH, this is where I'd really expect LLVM to do it's job (as in, I wouldn't consider optimising this much in MLIR). But your measurements might suggest you otherwise?

To me, it would make a lot of sense if you merged `extractNBitsFromVectorSigned` and `extractNBitsFromVectorUnsinged` - the only difference would be the right shift. Consistency is great :)

That said, I don't feel that strongly about it and I admit that it's a optimisation. If you decide to leave `arith::ShlOp + arith::AndIOp`, then could you add a comment in `extractNBitsFromVectorUnsinged` to document the reason for divergence? Something along the lines:

> NOTE: Similarly to `extractNBitsFromVectorUnsinged`, this could be used with arith::ShlOp + arith::ShrSOp`. However, by using `arith::ShlOp + arith::AndIOp`, we are eliminating shift left when the index is 0.

I know that this feels obvious today, but I always try to optimise for my future self who will forget :)

ziereis replied:
I have not measured if it would make any difference, i will look into this, for now i have just created the note. However changing this would also mean i have to change all the tests for the i4 to i8 rewrites.

ziereis commented on mlir/lib/Dialect/Vector/Transforms/VectorEmulateNarrowType.cpp:
```diff
@@ -1084,10 +1084,14 @@ static LogicalResult alignedConversionPrecondition(PatternRewriter &rewriter,
-  if (srcElemBitwidth != 4 || dstElemBitwidth < 8 ||
-      (dstElemBitwidth % srcElemBitwidth) != 0)
-    return rewriter.notifyMatchFailure(op, "Not a supported aligned case");
+  if (dstElemBitwidth < 8)
+    return rewriter.notifyMatchFailure(
+        op, "the bitwidth of dstType must be greater than or equal to 8");
+  if (dstElemBitwidth % srcElemBitwidth != 0)
+    return rewriter.notifyMatchFailure(op, "unaligned cases are not supported");
+  if (srcElemBitwidth != 2 && srcElemBitwidth != 4)
+    return rewriter.notifyMatchFailure(
+        op, "only src bitwidth of 2 or 4 is supported at this moment");
```
Thank you

ziereis committed:
update documentation

ziereis committed:
improve naming and comments

ziereis commented on mlir/lib/Dialect/Vector/Transforms/VectorEmulateNarrowType.cpp:
```diff
@@ -1172,70 +1176,144 @@ Value BitCastRewriter::genericRewriteStep(
-  i8VecShape.back() = i8VecShape.back() / i4Toi8BitwidthFactor;
-  auto i8VecType = VectorType::get(i8VecShape, rewriter.getI8Type());
-  Value i8Vector = rewriter.create<vector::BitCastOp>(loc, i8VecType, srcValue);
+/// Extracts a signed N-bit sequence from each element of an 8-bit vector,
+/// starting at the specified bit index.
```
> * for every byte `b` in `src` (which is a vector of bytes),
> * extracts `numBits` starting at `bitIdx` (let's call it `inputVal`), and
yes
> * returns a byte matching the value encoded in `inputVal`.

it will return a vector of bytes, the resultType will always be the same as the srcType. 
so for example lets say `numBits` is 4, it will treat the `inputVal` as a `i4` and (sign)ext it to a `i8` value.

banach-space replied:
`extractNbitsPerByteAndExtendToI8`?

Am I correct that this method assumes that the src and dst element type is `i8`?

ziereis replied:
yes

ziereis commented on mlir/lib/Dialect/Vector/Transforms/VectorEmulateNarrowType.cpp:
```diff
@@ -1172,70 +1176,144 @@ Value BitCastRewriter::genericRewriteStep(
+                                      Value srcValue) {
+  auto srcVecType = cast<VectorType>(srcValue.getType());
+  int64_t srcBitwidth = srcVecType.getElementType().getIntOrFloatBitWidth();
+  assert(8 % srcBitwidth == 0 && "Invalid source bitwidth");
```
yes it should be verified, i guess i can also remove this?

ziereis replied:
alright i will leave it.

ziereis committed:
improve naming and fix typo

banach-space reviewed (commented):
I love the updated comments 🙏🏻

ziereis committed:
tests for precondition and note for optimization

ziereis committed:
better name for extract function

banach-space commented on mlir/lib/Dialect/Vector/Transforms/VectorEmulateNarrowType.cpp:
```diff
@@ -1179,70 +1183,166 @@ Value BitCastRewriter::genericRewriteStep(
+/// Where aligned means it satisfies the alignedConversionPreconditions.
+///
+/// Example:
+/// vector<16x16xi2> -> vector<16x2xi8>
+/// vector<16x16xi4> -> vector<16x4xi8>
```
```suggestion
/// vector<16x16xi2> -> vector<16x4xi8>
/// vector<16x16xi4> -> vector<16x8xi8>
```

banach-space commented on mlir/lib/Dialect/Vector/Transforms/VectorEmulateNarrowType.cpp:
```diff
@@ -1179,70 +1183,166 @@ Value BitCastRewriter::genericRewriteStep(
+  int64_t srcBitwidth = srcVecType.getElementType().getIntOrFloatBitWidth();
+  assert(8 % srcBitwidth == 0 &&
+         "Unsupported sub-byte type (not a divisor of i8)");
+  int64_t bitwidthFactor = 8 / srcBitwidth;
```
To keep the naming consistent within the file, could this be `numSrcElemsPerByte`?

Similar variable elsewhere: https://github.com/llvm/llvm-project/blob/cfee344dda7394631f2177a15e56cfeee1d61fc4/mlir/lib/Dialect/Vector/Transforms/VectorEmulateNarrowType.cpp#L1098

ziereis replied:
done, there are some other places that refer to the same name but im not sure if they refer to the same thing

banach-space commented on mlir/lib/Dialect/Vector/Transforms/VectorEmulateNarrowType.cpp:
```diff
@@ -1179,70 +1183,166 @@ Value BitCastRewriter::genericRewriteStep(
-  i8VecShape.back() = i8VecShape.back() / i4Toi8BitwidthFactor;
-  auto i8VecType = VectorType::get(i8VecShape, rewriter.getI8Type());
-  Value i8Vector = rewriter.create<vector::BitCastOp>(loc, i8VecType, srcValue);
+/// Extracts a signed N-bit sequence from each element of an 8-bit vector,
```
[nit] This might be my "hardware" bias, but to me an 8-bit vector is literally a vector that holds 8-bits (for comparison, NEON has 128-bit vectors). 
```suggestion
/// Extracts a signed N-bit sequence from each element of a vector of bytes,
```

banach-space commented on mlir/lib/Dialect/Vector/Transforms/VectorEmulateNarrowType.cpp:
```diff
@@ -1179,70 +1183,166 @@ Value BitCastRewriter::genericRewriteStep(
+  // vec1  = [1,1,1,1],...
+  // vec2  = [2,2,2,2],...
+  // vec3  = [3,3,3,3],...
+  // 02    = [0,2,0,2,...],...
+  // 13    = [1,3,1,3,...],...
```
[nit] I think that this would help readability of this
```suggestion
  // 02    = [0,2,0,2,0,2,0,2],...
  // 13    = [1,3,1,3,1,3,1,3],...
```

banach-space commented on mlir/lib/Dialect/Vector/Transforms/VectorEmulateNarrowType.cpp:
```diff
@@ -1090,12 +1090,16 @@ static LogicalResult alignedConversionPrecondition(PatternRewriter &rewriter,
+        op, "only src bitwidth of 2 or 4 is supported at this moment");
 
-  const int numSrcElemsPerDestElem = dstElemBitwidth / srcElemBitwidth;
+  const int numSrcElemsPerDestElem = 8 / srcElemBitwidth;
```
Why hard-code this to 8?

Also, comment for L1105 (sadly GitHub doesn't allow comments for things outside the diff) - please update the error msg.

ziereis replied:
I changed this check since the dst in this case will always be a i8 vector.

For example:

%0 = arith.extui %a : vector<4xi2> to vector<4xi32>

should be possible because its only important that i can pack the 4 i2 values into a 1xi8 vector (otherwise i would have a invalid bitcast).

But i also added a test that checks something like this fails:

%0 = arith.extui %a : vector<2xi2> to vector<2xi32>

banach-space commented on mlir/test/Dialect/Vector/vector-rewrite-narrow-types.mlir:
```diff
@@ -193,6 +193,23 @@ func.func @f3ext(%a: vector<5xi8>) -> vector<8xi17> {
 }
 
+
+// CHECK-LABEL: func.func @aligned_i4_trailing_dim_not_multiple(
+func.func @aligned_i4_trailing_dim_not_multiple(%a: vector<1xi4>) -> vector<1xi8> {
+  // CHECK-NOT: arith.bitcast
+  // CHECK: arith.extsi %[[IN:.*]] : vector<1xi4> to vector<1xi8>
+  %0 = arith.extsi %a : vector<1xi4> to vector<1xi8>
+  return %0 : vector<1xi8>
+}
+
+// CHECK-LABEL: func.func @aligned_i2_trailing_dim_not_multiple(
+func.func @aligned_i2_trailing_dim_not_multiple(%a: vector<2xi2>) -> vector<2xi8> {
+  // CHECK-NOT: arith.bitcast
+  // CHECK: arith.extsi %[[IN:.*]] : vector<2xi2> to vector<2xi8>
+  %0 = arith.extsi %a : vector<2xi2> to vector<2xi8>
+  return %0 : vector<2xi8>
```
These are the unaligned cases, right? Please move to https://github.com/llvm/llvm-project/blob/main/mlir/test/Dialect/Vector/vector-emulate-narrow-type-unaligned.mlir.

banach-space replied:
Ping

ziereis replied:
This was meant like more of a negative test for the alignedConversionPrecondition. Also i think the file you linked refers to the void vector::populateVectorNarrowTypeEmulationPatterns and i changed the vector::populateVectorNarrowTypeRewritePatterns (in the same file) if i understood correctly?

banach-space replied:
Right, now I see. Then, still, `aligned` --> `unalagined` ;-) Also, to match the naming convention that follows this:
```mlir
@aligned_i4_trailing_dim_not_multiple --> @unaligned_extsi_i4_to_i8_trailing_dim_not_multiple
```

In fact, I'd do this:
```mlir
// Negative test - the trailing dim 2 is not a multiple of 4 (i.e. 8 / 2).

func.func @unaligned_extsi_i4_to_i8
```

Similar comment for the other negative test.

banach-space commented on mlir/test/Dialect/Vector/vector-rewrite-narrow-types.mlir:
```diff
@@ -206,6 +223,27 @@ func.func @aligned_extsi_i4_to_i8(%a: vector<8xi4>) -> vector<8xi8> {
+// CHECK:           %[[CST_4:.*]] = arith.constant dense<4> : vector<2xi8>
+// CHECK:           %[[CST_6:.*]] = arith.constant dense<6> : vector<2xi8>
+// CHECK:           %[[BITCAST:.*]] = vector.bitcast %[[IN]] : vector<8xi2> to vector<2xi8>
+// CHECK:           %[[SHL_6:.*]] = arith.shli %[[BITCAST]], %[[CST_6]] : vector<2xi8>
+// CHECK:           %[[ELEM0:.*]] = arith.shrsi %[[SHL_6]], %[[CST_6]] : vector<2xi8>
+// CHECK:           %[[SHL_4:.*]] = arith.shli %[[BITCAST]], %[[CST_4]] : vector<2xi8>
+// CHECK:           %[[ELEM1:.*]] = arith.shrsi %[[SHL_4]], %[[CST_6]] : vector<2xi8>
+// CHECK:           %[[SHL_2:.*]] = arith.shli %[[BITCAST]], %[[CST_2]] : vector<2xi8>
+// CHECK:           %[[ELEM2:.*]] = arith.shrsi %[[SHL_2]], %[[CST_6]] : vector<2xi8>
+// CHECK:           %[[ELEM3:.*]] = arith.shrsi %[[BITCAST]], %[[CST_6]] : vector<2xi8>
```
[nit] To improve readability (same comment for other tests)

```suggestion
// Extract bits 0-1
// CHECK:           %[[SHL_6:.*]] = arith.shli %[[BITCAST]], %[[CST_6]] : vector<2xi8>
// CHECK:           %[[ELEM0:.*]] = arith.shrsi %[[SHL_6]], %[[CST_6]] : vector<2xi8>
// Extract bits 2-3
// CHECK:           %[[SHL_4:.*]] = arith.shli %[[BITCAST]], %[[CST_4]] : vector<2xi8>
// CHECK:           %[[ELEM1:.*]] = arith.shrsi %[[SHL_4]], %[[CST_6]] : vector<2xi8>
// Extract bits 4-5
// CHECK:           %[[SHL_2:.*]] = arith.shli %[[BITCAST]], %[[CST_2]] : vector<2xi8>
// CHECK:           %[[ELEM2:.*]] = arith.shrsi %[[SHL_2]], %[[CST_6]] : vector<2xi8>
// Extract bits 6-7
// CHECK:           %[[ELEM3:.*]] = arith.shrsi %[[BITCAST]], %[[CST_6]] : vector<2xi8>
```

banach-space reviewed (commented):
Thanks for all the updates, this is looking really neat 🙏🏻  A few final comments, nothing major.

ziereis committed:
Update mlir/lib/Dialect/Vector/Transforms/VectorEmulateNarrowType.cpp

Co-authored-by: Andrzej Warzyński <andrzej.warzynski@gmail.com>

ziereis committed:
Update mlir/lib/Dialect/Vector/Transforms/VectorEmulateNarrowType.cpp

Co-authored-by: Andrzej Warzyński <andrzej.warzynski@gmail.com>

ziereis committed:
Update mlir/lib/Dialect/Vector/Transforms/VectorEmulateNarrowType.cpp

Co-authored-by: Andrzej Warzyński <andrzej.warzynski@gmail.com>

ziereis committed:
better naming and comments for test

ziereis commented:
Thanks again for the comments

hanhanW reviewed (approved):
I like the updated documentation, thanks a lot! I just have a final optional nit.

ziereis committed:
move assertion

ziereis commented:
@banach-space Sorry, I didn’t mean to ignore the comments. I somehow started my own review on the PR or something, so all the comments were only "pending." I didn’t know what this meant, but it appears they were only visible to me. 😅 I hope they’re visible now!

banach-space commented on mlir/lib/Dialect/Vector/Transforms/VectorEmulateNarrowType.cpp:
```diff
@@ -1090,15 +1090,20 @@ static LogicalResult alignedConversionPrecondition(PatternRewriter &rewriter,
+  if ((subByteVecType.getShape().back() % numSrcElemsPerByte) != 0)
     return rewriter.notifyMatchFailure(
-        op, "Not an even number of i4 elements in trailing dim");
+        op, "the trailing dimension of subByteVecType must be a multiple of "
+            "the srcElemBitwidth / 8");
```
This doesn't feel right: "a multiple of the srcElemBitwidth / 8". Did you mean "a multiple of "8 / the srcElemBitwidth"?

In general, this is quite tricky to describe in words. Here's a suggestion:

```suggestion
        op, "the trailing dimension of the input vector of sub-bytes must be a multiple of "
            "8 / <sub-byte-width>");
```

banach-space commented on mlir/lib/Dialect/Vector/Transforms/VectorEmulateNarrowType.cpp:
```diff
@@ -1179,70 +1184,166 @@ Value BitCastRewriter::genericRewriteStep(
 
-  // 3. Interleave low and high i8 elements.
-  return rewriter.create<vector::InterleaveOp>(loc, low, high);
+/// Extracts an unsigned N-bit sequence from each element of an 8-bit vector,
```
Simiarly to `extractNBitsPerByteAndSignExtendToI8`, please replace `8-bit vector` with `a vector of bytes" ;-)

banach-space reviewed (approved):
LGTM % some minor requests.

This is quite intricate, and I really appreciate the effort you've put into implementing this and so patiently addressing all my comments.

ziereis committed:
Update mlir/lib/Dialect/Vector/Transforms/VectorEmulateNarrowType.cpp

Co-authored-by: Andrzej Warzyński <andrzej.warzynski@gmail.com>

ziereis committed:
better naming and comments for test

ziereis commented:
The comments should be resolved now. Thanks as well for all the reviews and being patient with me :)

banach-space commented:
> Thanks as well for all the reviews and being patient with me :)

Works both ways :) Do you have commit access?

ziereis committed:
Merge branch 'main' into i2toi8-ext-rewrites

ziereis commented:
no, i would be happy if you could merge it.

banach-space commented:
> no, i would be happy if you could merge it.

Landed 🥳 Thanks for working on this 🙏🏻  

Note, GitHub added me as a co-author. AFAIK, that's the default behaviour when squashing commits like this:
* https://github.com/llvm/llvm-project/pull/121298/commits/8abb46d6418762be0609c87927c22e2a392153b1

(IIUC, you've incorporated my suggestion through GitHub's UI). I don't really mind, but obviously you've done 99% of the work here 😅
